client/verta/verta/deployment/_deployedmodel.py

```diff
@@ -132,46 +117,52 @@ def prediction_url(self, value):
             raise ValueError("not a valid prediction_url")
         self._prediction_url = parsed.geturl()
 
-    @property
-    def credentials(self):
-        return self._credentials
-
-    @credentials.setter
-    def credentials(self, value):
-        self._credentials = value
-        self._init_session()
-
-    @property
-    def access_token(self):
-        return self._access_token
 
-    @access_token.setter
-    def access_token(self, value):
-        self._access_token = value
-        self._init_session()
+    # TODO: Implement dynamic compression via separate utility and call it from here
+    def _predict(
+            self,
+            x: Any,
+            compress: bool=False,
+            prediction_id: Optional[str]=None,
+            ):
+        """This is like ``DeployedModel.predict()``, but returns the raw ``Response``
+         for debugging."""
+        request_headers = dict()
+        if prediction_id:
+            request_headers.update({'verta-request-id': prediction_id})
 
-    def _predict(self, x, compress=False):
-        """This is like ``DeployedModel.predict()``, but returns the raw ``Response`` for debugging."""
         x = _utils.to_builtin(x)
         if compress:
+            request_headers.update({'Content-Encoding': 'gzip'})
             # create gzip
             gzstream = six.BytesIO()
             with gzip.GzipFile(fileobj=gzstream, mode="wb") as gzf:
                 gzf.write(six.ensure_binary(json.dumps(x)))
             gzstream.seek(0)
 
-            return self._session.post(
+            response = self._session.post(
                 self._prediction_url,
-                headers={"Content-Encoding": "gzip"},
+                headers=request_headers,
                 data=gzstream.read(),
-            )
+                )
         else:
-            return self._session.post(self.prediction_url, json=x)
+            response = self._session.post(
+                self.prediction_url,
+                headers=request_headers,
+                json=x,
+                )
+        if not response.ok:
+            _utils.raise_for_http_error(
+                response=response,
+                component_msg='Deployed model encountered an error: ')
```

Comment:
This is not quite what I was trying to indicate with my suggestion https://github.com/VertaAI/modeldb/pull/3332#discussion_r1014483505

- This behavior should only be for `400`s. This would apply to all error codes, but would not be accurate to prepend other error codes with "Deployed model encountered an error".
- I'd prefer if we preserved the old behavior of raising a `RuntimeError` for `400`s, rather than changing it to an `HTTPError`, for compatibility with existing user code that might be trying to catch it, but I'm not wholly opposed to this.
- This approach renders a massive stack trace that exposes all of our middleware, and users tend to have a hard time interpreting large stack traces (not to mention the logging verbosity if this is thrown in a microservice):
  ```python
  ---------------------------------------------------------------------------
  HTTPError                                 Traceback (most recent call last)
  Cell In [3], line 2
        1 data = [1, 2, 3]
  ----> 2 assert endpoint.get_deployed_model().predict(data) == data
  
  File ~/Documents/modeldb/client/verta/verta/deployment/_deployedmodel.py:255, in DeployedModel.predict(self, x, compress, max_retries, always_retry_404, always_retry_429, retry_status, backoff_factor, prediction_id)
      247 if always_retry_404 or always_retry_429:
      248     warnings.warn(
      249         "Deprecation warning: The `always_retry_404` and `always_retry_429`"
      250         "parameters are deprecated and will removed in an upcoming version; "
     (...)
      253         category=FutureWarning,
      254     )
  --> 255 prediction_with_id: Tuple[str, Any] = self.predict_with_id(
      256     x=x,
      257     compress=compress,
      258     prediction_id=prediction_id,
      259     max_retries=max_retries,
      260     retry_status=retry_status,
      261     backoff_factor=backoff_factor,
      262     )
      263 return prediction_with_id[1]
  
  File ~/Documents/modeldb/client/verta/verta/deployment/_deployedmodel.py:325, in DeployedModel.predict_with_id(self, x, compress, max_retries, retry_status, backoff_factor, prediction_id)
      317 # Set the retry config if it differs from current config.
      318 self._session = http_session.set_retry_config(
      319     self._session,
      320     max_retries=max_retries,
      321     status_forcelist=retry_status,
      322     backoff_factor=backoff_factor,
      323     )
  --> 325 response = self._predict(x, compress, prediction_id)
      326 id = response.headers['verta-request-id']
      327 return (id, _utils.body_to_json(response))
  
  File ~/Documents/modeldb/client/verta/verta/deployment/_deployedmodel.py:155, in DeployedModel._predict(self, x, compress, prediction_id)
      149     response = self._session.post(
      150         self.prediction_url,
      151         headers=request_headers,
      152         json=x,
      153         )
      154 if not response.ok:
  --> 155     _utils.raise_for_http_error(
      156         response=response,
      157         component_msg='Deployed model encountered an error: ')
      158 return response
  
  File ~/Documents/modeldb/client/verta/verta/_internal_utils/_utils.py:596, in raise_for_http_error(response, component_msg)
      593 message = f"{component_msg or ''}{response.status_code} {cause} Error:" \
      594           f" {reason} for url: {response.url}"
      595 message += time_str  # attach time to error message
  --> 596 six.raise_from(requests.HTTPError(message, response=response), None)
  
  File <string>:3, in raise_from(value, from_value)
  
  HTTPError: Deployed model encountered an error: 400 Client Error: Traceback (most recent call last):
    File "/Users/miliu/Documents/modeldb/client/verta/verta/registry/_verify_io.py", line 76, in wrapper
    File "/var/folders/_b/kwcscwf12kl4213t2kwdctq00000gn/T/ipykernel_54729/3205215150.py", line 10, in predict
  KeyError: 'non existent key'
   for url: https://app.verta.ai/api/v1/predict/simple-deploy at 2022-11-16 18:59:01.442000 UTC
  ```
  I'd recommend either finding a way to suppress the inner frames, or using the old approach I linked to in my original suggestion and we can revisit this later. I think grabbing `raise_for_http_error()` is a good idea, but it'd need some more work.